File: bitmaps/src/PreviewBitmap.cpp
#include "bitmaps.h"

#include <string.h>

void PreviewBitmap_Init(PreviewBitmap* bmp, int width, int height)
{
	bmp->width = width;
	bmp->height = height;

	bmp->r = new Int8[width * height];
	bmp->g = new Int8[width * height];
	bmp->b = new Int8[width * height];
}

void PreviewBitmap_Copy(PreviewBitmap* src, PreviewBitmap* res)
{
	PreviewBitmap_Init(res, src->width, src->height);

	memcpy(res->r, src->r, src->width * src->height * sizeof(Int8));
	memcpy(res->g, src->g, src->width * src->height * sizeof(Int8));
	memcpy(res->b, src->b, src->width * src->height * sizeof(Int8));
}

void PreviewBitmap_Free(PreviewBitmap* fb)
{
	delete [] fb->r;
	delete [] fb->g;
	delete [] fb->b;
}

void PreviewBitmap_FromPreciseBitmap(PreciseBitmap* src, PreviewBitmap* res)
{
	PreviewBitmap_Init(res, src->width, src->height);

	for (int k = 0; k < src->width * src->height; k++)
	{
		res->r[k] = (Int8)(src->r[k] * 255);
		res->g[k] = (Int8)(src->g[k] * 255);
		res->b[k] = (Int8)(src->b[k] * 255);
	}
}

File: CatEyeB.Core/com/cateye/core/native_/PreviewBitmap.java
package com.cateye.core.native_;

import com.cateye.core.IPreviewBitmap;

class PreviewBitmap implements IPreviewBitmap
{
	/**
	 * Width of the image
	 */
	int width;
	
	/**
	 * Height of the image
	 */
	int height;
	
	/**
	 * Pointer to the red channel
	 */
	long r;
	
	/**
	 * Pointer to the green channel
	 */
	long g;
	
	/**
	 * Pointer to the blue channel
	 */
	long b;
	
	@Override
	public int getWidth()
	{
		return width;
	}
	
	@Override
	public int getHeight()
	{
		return height;
	}
	
	@Override
	public void dispose()
	{
		checkResult(Free(this));
	}
	
	@Override
	public IPreviewBitmap clone()
	{
		PreviewBitmap result = new PreviewBitmap();
		checkResult(Copy(this, result));
		
		return result;
	}
	
	/**
	 * Check result and if it is wrong throw the exception
	 */
	private void checkResult(int resultCode)
	{
		switch (resultCode)
		{
			case BITMAP_RESULT_OUT_OF_MEMORY:
				throw new OutOfMemoryError();
			case BITMAP_RESULT_INCORRECT_DATA:
				throw new IncorrectDataException();
		}
	}
	
	static final int BITMAP_RESULT_OK = 0;
	static final int BITMAP_RESULT_OUT_OF_MEMORY = 1;
	static final int BITMAP_RESULT_INCORRECT_DATA = 2;

	static native int Init(PreviewBitmap bmp, int width, int height);
	static native int Copy(PreviewBitmap src, PreviewBitmap res);
	static native int Free(PreviewBitmap fb);
	
	static
	{
		LibraryLoader.attach("bitmaps.dll");
	}
}
